### PR TITLE
codeintel: Auto-inference sandbox API

### LINKED ISSUE #33041: RFC 624: (M1) Re-build PoC recognizers
Implements a portion of [RFC 624: Configurable auto-indexing recognizers](https://docs.google.com/document/d/1LOVI6xogkt4Iifl1F_8Gb_H9pXQ_MrynW8vmr6D05A0).

Re-implement the [existing recognizer behaviors](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cd2182763d2839b43cd49edfc877aa2ac5a07b80/-/blob/lib/codeintel/autoindex/inference/recognizers.go?L23:5#tab=references) in Lua using the exact interfaces that are declared today. Create a version of a Lua sandbox introduced in #33042 that exposes the gitserver client methods to the script. Ensure that the current recognizer unit tests pass against this new sandbox executor.

This issue does not yet make the sandbox accessible outside of tests.

### PR DESCRIPTION
**Pull request by efritz**
This PR adds an initial `internal/codeintel/autoindexing/internal/inference` package that encodes the rules of our autoindexing inference in a Lua sandbox. This package is currently unused by any entrypoint (outside of unit tests) and subsequent efforts (#33044 and #33045) will replace this code with the original.

Closes #33041 and #33040.

This PR does not close discussion on API design. This PR has roughed-out what seems to be a very usable interface for us to be able to extend. We're still waiting on customer feedback (see #33047) to help us ensure we're solving the right expressibility problem, but it doesn't seem to be a blocker for moving the PoC forward (but may be a blocker before moving onto or past #33046).

**Reviewer hints**:

Please review by-commit, paying special attention to the following:

- 766e2547 adds the entire service implementation
- e1894e18 reimplements the logic of [`lib/codeintel/autoindex`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@0b182d47000bbbc488463995d577875b99baf826/-/tree/lib/codeintel/autoindex) in Lua
- 7eac9d00 adds unit tests

## Test plan

This PR adds significantly good coverage and shows that we have one-to-one behavior with our existing indexers (test cases are taken verbatim). Additional testing will be done before this code is hooked up to a production (or even development application) entrypoint.

### CODE REVIEW

File: internal/luasandbox/codeintel/lua/recognizers.lua [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+})
+
+local gomod_recognizer = sgrecognizer.pathRecognizer {
+    patterns = {
+        sgpattern.basename("go.mod"),
+        sgpattern.exclude(go_exclude_paths),
+    },
+
+    generate = function(api, paths)
+        local image = "sourcegraph/lsif-go:latest"
```
Comment (tjdevries) [inline internal/luasandbox/codeintel/lua/recognizers.lua:41]: could make `sg.images` to store these, so ppl don't typo strings.

Comment (efritz) [reply internal/luasandbox/codeintel/lua/recognizers.lua:41]: This is just to get 1-1 with current behavior

File: internal/luasandbox/codeintel/lua/recognizers.lua [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                    outfile = "",
+                    root = "",
+                    steps = {},
+                }
+            end
+        end
+    end,
+}
+
+local go_recognizer = sgrecognizer.fallbackRecognizer {
```
Comment (tjdevries) [inline internal/luasandbox/codeintel/lua/recognizers.lua:89]: one thing that feels weird about `fallbackRecognizer` is you specify fallbacks but no `primary` (explicitly, from looking at the other code, looks fine. just from user API perspective).

Maybe should make a recognizer and have an `addFallback(...)` method or something

File: internal/codeintel/autoindexing/inference/lua/util.lua [hunk 3]
```diff
@@ -0,0 +1,44 @@
+local contains = function(table, element)
```
Comment (efritz) [inline internal/codeintel/autoindexing/inference/lua/util.lua:1]: Note: All of this is just a minimal v1 to get the behavior we have today. @tjdevries suggested something like [penlight](https://github.com/lunarmodules/Penlight) as a stdlib and I'm completely for that (but don't want to bikeshed too hard on the _lack of polish_ on the PR).

If there are particular ergonomic suggestions, please present them as a "this is how the user should expression this" and we can work backwards from there. Note these will be handled post-merge of this initial PR.

**Comment by mrnugget** [comment]
High-level question: the PR description says this closes https://github.com/sourcegraph/sourcegraph/issues/33041 and https://github.com/sourcegraph/sourcegraph/issues/33040 which belong to milestones 1 and 4 respectively. [This comment in the RFC](https://docs.google.com/document/d/1LOVI6xogkt4Iifl1F_8Gb_H9pXQ_MrynW8vmr6D05A0/edit?disco=AAAAWTAmH8c) makes it sound like we still haven't finished milestone 0 yet.

So why do this now? Did we actually have to implement this to get some feedback from customers? Because if we haven't heard from customers yet on RFC 624 and whether it'll solve their problem, this is a lot of code to build on a hunch. What I'm thinking is: could we have written pseudo Lua code and asked them whether they'd use it and whether that would work for their codebase?

**Comment by efritz** [comment]
High-level response: what happened in the week while @mrnugget was OoO and some misc thoughts:

- We've reached out to four customers this iteration to adopt auto-indexing. One of them is blocked on their internal Java team. One of them uses primarily Go, so we're all set for them. The other two we've asked specific questions about their repository setup and how they would auto-indexing their code by hand. They've both responded this morning so we'll be following up to gather more useful information.
- Increasing momentum on people setting up auto-indexing and running into the "what now" part when it comes to getting it to configure jobs is putting pressure on having _some_ sort of inference-logic-as-configuration solution. The majority of RFC 624 is a direct solution to this (with some additional work on exporting it for users as polished interface). I feel that M0 (API validation) is a blocker for making it a user-facing feature, but not to do the technical work up front to give us the benefit of modifying a running instance without having to hard-code customer-specific rules in the next patch release.
- This PR combines parts of M1 (Lua port of `lib/codeintel/autoindexing/inference` and M4 (update/optimize gitserver communication) as it turns out it would have been more total work to do a _direct_ port in M1 and then change it later for M4. @tjdevries gave a lot of insights to help inform these decisions as well.

What we should do going forward:

- Land M1 and M4 (this PR)
- Add additional rate limiting and observation around gitserver calls from Lua (M2: harden gitserver client). We've made the number of requests more efficient across all recognizers, but we're still changing the exact set of things we're running and can't be too careful.
- Add a backdoor way to insert configuration into an instance so that the next release will have some mid-release way to say "Ooh you want support for `pom.xml.rft`? No problem!" and can give them a string to insert directly into their database for now.
- Later (not urgently) polish this and advertise it as a possibility for users (not an ace-in-our-sleeve during experiments).

I'm very confident that the solution here is not far off from the solution we'd get if we blocked on completion of M0. We have enough uses cases we don't support today for Java alone (no build tooling support today, only package repos) that the API in this PR is perfectly suited for.

Completion of M0 is still super important because the _Lua sandbox in this PR is not yet ergonomic_. It's a bare-bones v1 (that improves on the existing art), but it's not idiomatic and has no standard library attached yet. (I'll be deferring heavily to customer response and @tjdevries's feedback here).

**Review by olafurpg — APPROVED** [review]
LGTM 👍 I like the new lua inference logic and it will be a gamechanger for users to be able to roll out custom inference changes.

File: internal/codeintel/autoindexing/inference/service_test.go [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver"
+	"github.com/sourcegraph/sourcegraph/internal/luasandbox"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+	"github.com/sourcegraph/sourcegraph/internal/unpack/unpacktest"
+	"github.com/sourcegraph/sourcegraph/lib/codeintel/autoindex/config"
+)
+
+func TestRecognizers(t *testing.T) {
+	testCases := []struct {
```
[the remark this answers is not in the document]
Comment (efritz) [reply internal/codeintel/autoindexing/inference/service_test.go:22]: Sure, i can break these up and just make a shared helper function no problem.

File: internal/codeintel/autoindexing/inference/libs.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	lua "github.com/yuin/gopher-lua"
+
+	"github.com/sourcegraph/sourcegraph/internal/codeintel/autoindexing/inference/libs"
+	"github.com/sourcegraph/sourcegraph/internal/luasandbox"
+	"github.com/sourcegraph/sourcegraph/internal/luasandbox/util"
+)
+
+var defaultAPIs = map[string]luasandbox.LuaLib{
+	"sg.path_patterns": libs.Patterns,
```
Comment (tjdevries) [inline internal/codeintel/autoindexing/inference/libs.go:12]: if we're going to have patterns that aren't for paths, I'd probably rather see `sg.patterns.path` or we just call this `sg.pattern` and then have `path_` prefixes on them or something

File: internal/codeintel/autoindexing/inference/lua/go.lua [hunk 6]
```diff
@@ -0,0 +1,73 @@
+local path = require("path")
+local path_patterns = require("sg.path_patterns")
+local recognizers = require("sg.recognizers")
+
+local shared = loadfile("shared.lua")()
```
Comment (tjdevries) [inline internal/codeintel/autoindexing/inference/lua/go.lua:5]: any reason we can't add `shared.lua` to the package.path  or something so you can just `require('shared')`

File: internal/codeintel/autoindexing/inference/lua/go.lua [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+local gomod_recognizer = recognizers.path_recognizer {
+    patterns = {
+        path_patterns.basename("go.mod"),
+        path_patterns.exclude(exclude_paths),
+    },
+
+    -- Invoked when go.mod files exist
+    generate = function(api, paths)
+        local jobs = {}
+        for i = 1, #paths do
```
Comment (tjdevries) [inline internal/codeintel/autoindexing/inference/lua/go.lua:22]: we're doing the transform stuff for paths later then?

File: internal/codeintel/autoindexing/inference/lua/typescript.lua [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    end
+
+    return false
+end
+
+local infer_typescript_job = function(api, tsconfig_path, should_infer_config)
+    local root = path.dirname(tsconfig_path)
+    local reverse_ancestors = util.reverse(path.ancestors(tsconfig_path))
+
+    api:callback(recognizers.path_recognizer {
```
Comment (tjdevries) [inline internal/codeintel/autoindexing/inference/lua/typescript.lua:57]: I don't remember why we have this `:callback` thing

File: internal/codeintel/autoindexing/inference/lua/util.lua (hunk 3 above)
Comment (tjdevries) [reply internal/codeintel/autoindexing/inference/lua/util.lua:1]: penlight is good choice probably. Depends what areas we're trying to cover though. Can chat later about it.

File: internal/codeintel/autoindexing/inference/lua/go.lua (hunk 6 above)
Comment (efritz) [reply internal/codeintel/autoindexing/inference/lua/go.lua:5]: https://github.com/sourcegraph/sourcegraph/issues/34295

File: internal/codeintel/autoindexing/inference/lua/typescript.lua (hunk 8 above)
Comment (efritz) [reply internal/codeintel/autoindexing/inference/lua/typescript.lua:57]: We need _additional_ information from gitserver once we've already hit a recognizer. It's basically registering that path recognizer as a thing to invoke next (like the JS event loop).

File: internal/codeintel/autoindexing/inference/lua/util.lua (hunk 3 above)
Comment (efritz) [reply internal/codeintel/autoindexing/inference/lua/util.lua:1]: https://github.com/sourcegraph/sourcegraph/issues/34297

File: internal/codeintel/autoindexing/inference/lua/typescript.lua (hunk 8 above)
Comment (tjdevries) [reply internal/codeintel/autoindexing/inference/lua/typescript.lua:57]: ah. `callback` just seems a bit too generic. Is there a better name? (not sure I can think of one)

Comment (efritz) [reply internal/codeintel/autoindexing/inference/lua/typescript.lua:57]: `api:callback(...)` -> `recognizers.register(...)`?

File: internal/codeintel/autoindexing/inference/service_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (efritz) [reply internal/codeintel/autoindexing/inference/service_test.go:22]: 5ecea138f9b378306882e3f81e9d873437d4c8e6

[the remark this answers is not in the document]
Comment (efritz) [reply internal/codeintel/autoindexing/inference/service_test.go]: 5ecea138f9b378306882e3f81e9d873437d4c8e6

File: internal/codeintel/autoindexing/inference/libs.go (hunk 5 above)
Comment (efritz) [reply internal/codeintel/autoindexing/inference/libs.go]: 0faff06d9e2810d501306ef75e920feb1a9ef55c

File: internal/codeintel/autoindexing/inference/lua/typescript.lua (hunk 8 above)
Comment (efritz) [reply internal/codeintel/autoindexing/inference/lua/typescript.lua]: I think this helps a little bit - we can polish later as well. 

90eb1f7cf75a83ae2f0b9752f9b0a664cc8b8200

**Comment by efritz** [comment]
Enabling auto-merge now, but would be very happy to receive another review. We'll be iterating on this in the very near future.

**Pull request merged**